### PR TITLE
Add UnivariateTimeTypeToContinuous transformer to builtins

### PR DESCRIPTION
**Pull request by venuur**
Implements univariate case of TimeType transformer discussed in https://github.com/alan-turing-institute/MLJModels.jl/issues/234

I would appreciate feedback, especially recommendations on how to add tests.

### CODE REVIEW

File: src/builtins/Transformers.jl [hunk 1]
```diff
@@ -326,6 +326,56 @@ end
...
+    zero_time::Union{Nothing, TimeType}
+    step::Period
+    function UnivariateTimeTypeToContinuous(;
+        zero_time=nothing, step=Dates.Hour(24))
+        return new(zero_time, step)
+    end
+end
+
+function MLJBase.fit(model::UnivariateTimeTypeToContinuous, verbosity::Int, X)
+    if model.zero_time !== nothing
```
Comment (ablaom) [inline src/builtins/Transformers.jl:352]: Add check that zero_time has type compatible with actual type encountered.

File: src/builtins/Transformers.jl [hunk 2]
```diff
@@ -326,6 +326,56 @@ end
...
+        min_dt = minimum(X)
+        fitresult = min_dt
+    end
+    cache = nothing
+    report = nothing
+    return fitresult, cache, report
+end
+
+function MLJBase.transform(model::UnivariateTimeTypeToContinuous, fitresult, X)
+    # Set the size of a single step.
```
Comment (ablaom) [inline src/builtins/Transformers.jl:364]: Add check that `X` has the same type of vector encountered in fit.

File: src/builtins/Transformers.jl [hunk 3]
```diff
@@ -326,6 +326,56 @@ end
 MLJBase.inverse_transform(transformer::UnivariateStandardizer, fitresult, w) =
     [inverse_transform(transformer, fitresult, y) for y in w]
 
+##########################################################################################
+## CONTINUOUS TRANSFORM OF TIME TYPE FEATURES
+
+"""
+    UnivariateTimeTypeToContinuous(zero_time=nothing, step=Hour(24))
+
+Convert `Date`, `DateTime`, and `Time` vectors to `Float64` by assuming `0.0` corresponds
```
Comment (ablaom) [inline src/builtins/Transformers.jl:335]: Minor: I suggest "Convert **a** `Date`, `DateTime`, and `Time` **vector** to prevent confusion with multi-variate case.

**Comment by venuur** [comment]
Good feedback. I’ll update the PR soon.

**Comment by venuur** [comment]
I added the checks and I pushed the new version to the dev branch of venuur/MLJModels.jl. According to github the changes should show up here (though I don't see them yet). I'll be adding tests later in the week.

**Comment by ablaom** [comment]
> According to github the changes should show up here (though I don't see them yet). That's weird. 

This latest commit here is 8881832, which matches the latest commit on your dev branch just now:

Screen Shot 2020-05-06 at 11 25 04 AM

Are you sure they are not there?

**Comment by venuur** [comment]
When I look at the commits tab up top, I see the commits in the attached screenshot. 
3BE41D16-B29F-4C75-9551-38F5D34B2A21

The date doesn’t match what you are showing.

**Comment by ablaom** [comment]
Your latest commit appears there, it seems to me.  It's at the bottom. Perhaps what's confusing you is that this tab lists the commits in the opposite order to that in the conversations tab?

Sorry I can't be more helpful.

**Comment by venuur** [comment]
My latest commit on my fork is https://github.com/venuur/MLJModels.jl/commit/3abaa0965d299bd22dfaff0302c781357270d082

That is not the one on this pull request, right? Or am I just reading it all wrong, which is totally possible.

**Comment by ablaom** [comment]
No sorry, I must have been looking in the wrong place. This is quite weird. Going to close and re-open - see if that helps

**Comment by ablaom** [comment]
Nah. Can you add `# testing push` at the top line of src/MLJBase.jl, commit and push, and let's see if that appears here.

**Comment by venuur** [comment]
Looks like it's updated finally.

File: src/builtins/Transformers.jl [hunk 4]
```diff
@@ -342,15 +342,61 @@ called.
...
-        zero_time=nothing, step=Dates.Hour(24))
-        return new(zero_time, step)
+end
+
+function UnivariateTimeTypeToContinuous(;
+    zero_time=nothing, step=Dates.Hour(24))
+    model = UnivariateTimeTypeToContinuous(zero_time, step)
+    MLJBase.clean!(model)
+    return return model
+end
```
Comment (ablaom) [inline src/builtins/Transformers.jl:352]: You need to return a warning if `clean!` generates one:
```julia
    message = MMI.clean!(model)
    isempty(message) || @warn message
```julia

File: src/builtins/Transformers.jl [hunk 5]
```diff
@@ -342,15 +342,61 @@ called.
...
+                    model.step = convert(Hour, model.step)
+                else
+                    # Unable to resolve, rethrow original error.
+                    throw(err)
+                end
+            else
+                throw(err)
+            end
+        end
     end
```
Comment (ablaom) [inline src/builtins/Transformers.jl:382]: Small misunderstanding here: `clean!` does not actually issue the warnings, just appends them to it's `message` return value (initialised to "" at the beginning). Please see the example at  https://alan-turing-institute.github.io/MLJ.jl/dev/adding_models_for_general_use/#New-model-type-declarations-and-optional-clean!-method-1 .

Comment (ablaom) [reply src/builtins/Transformers.jl:382]: Finicky point: When referring to fields in the warnings, put in quotes, as is in "The value of `lambda` is not valid".

**Comment by ablaom** [comment]
Looking great! Thanks adding such detailed checks. 

I think all that's needed now are some tests.

**Comment by ablaom** [comment]
@venuur Be great to have some tests when you get a chance.

**Comment by venuur** [comment]
Thanks for pinging me. I haven’t forgotten about this, it’s just my workload has been heavier than usual so it may be another week or two before I can wrap this up. 

I’ll definitely get to it, the timeline is just a bit pushed compared to my original plan.

**Comment by ablaom** [comment]
No worries at all!!

**Comment by venuur** [comment]
Added tests and fixed a couple corner cases while writing the log tests. There's an annoying issue that the default constructor uses a step of Hour(24), but when applied to a Date, we need to convert to Day(1), which is fine, but since the constructor doesn't distinguish a user supplied step versus the default, a warning message is printed. I would rather add some extra checks to avoid the message if it's the default value, but I don't have the time right now to figure out the right logic. 

Excepting the somewhat unnecessary log message, the current version should be complete. Please let me know if I'm missing anything.

File: src/builtins/Transformers.jl (hunk 5 above)
Comment (venuur) [reply src/builtins/Transformers.jl:342]: Fixed the `clean!` method semantics, but forgot the backticks. If I get time I'll revisit, but not sure when that will be.

**Comment by venuur** [comment]
Should be all good for your next review when you get a chance @ablaom

**Comment by ablaom** [comment]
And, if you haven't noticed, there's a failure for julia 1.0. Here's an excerpt:

```
┌ Warning: Cannot add `TimePeriod` `step` to `Date` `zero_time`. Converting `step` to `Day`.
453└ @ MLJModels ~/build/alan-turing-institute/MLJModels.jl/src/builtins/Transformers.jl:408
454TimeTypeToContinuous: Test Failed at /home/travis/build/alan-turing-institute/MLJModels.jl/test/builtins/Transformers.jl:125
455  Expression: all(dt_continuous .== ex)
456Stacktrace:
457 [1] macro expansion at /home/travis/build/alan-turing-institute/MLJModels.jl/test/builtins/Transformers.jl:125 [inlined]
458 [2] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
459 [3] top-level scope at /home/travis/build/alan-turing-institute/MLJModels.jl/test/builtins/Transformers.jl:96
460TimeTypeToContinuous: Test Failed at /home/travis/build/alan-turing-institute/MLJModels.jl/test/builtins/Transformers.jl:134
461  Expression: all(dt_continuous .== ex)
462Stacktrace:
463 [1] macro expansion at /home/travis/build/alan-turing-institute/MLJModels.jl/test/builtins/Transformers.jl:134 [inlined]
464 [2] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.0/Test/src/Test.jl:1083 [inlined]
465 [3] top-level scope at /home/travis/build/alan-turing-institute/MLJModels.jl/test/builtins/Transformers.jl:96
466┌
```

**Comment by venuur** [comment]
Sounds good. I’ll make the changes you suggested and look into the Julia 1.0 error.

**Comment by ablaom** [comment]
Okay, working on the conflicts and adding the extras referred to above ...

**Comment by venuur** [comment]
Thanks for following up. And sorry for disappearing. As you may guess, I haven’t had time and won’t have time for a while to work on this. I’m still happy to comment here where I can.

**Comment by ablaom** [comment]
No worries. You put a lot a fine work into this. 

Note the multivariate (table) case will be handled by https://github.com/alan-turing-institute/MLJModels.jl/issues/288 . 

Closing in favour of #295.
